src/lj_api.c

```diff
@@ -752,6 +752,59 @@ LUA_API void *lua_newuserdata(lua_State *L, size_t size)
   return uddata(ud);
 }
 
+#if LJ_HASFFI
+#include "lj_cdata.h"
+
+LUA_API void *lua_newcdata(lua_State *L, int ctid, size_t size)
+{
+  CTState *cts = ctype_cts(L);
+  GCcdata *cd = lj_cdata_new(cts, ctid, size);
+  setcdataV(L, L->top, cd);
+  incr_top(L);
+  lj_gc_check(L);
+  return cdataptr(cd);
+}
+
+LUA_API void lua_pushcdataptr(lua_State *L, const void *ptr)
+{
+  const void **cd = lua_newcdata(L, CTID_P_CVOID, sizeof(ptr));
+  *cd = ptr;
+}
+
+#include "lj_cparse.h"
+LUA_API int lua_getctid(lua_State *L, int narg)
+{
+  cTValue *o = index2adr(L, narg);
+
+  if (tvisnumber(o)) {
```

Comment:
I'm not sure this branch is the best idea. e.g. what happens when a user passes in an int64?